architecture-docs/adr-003.md

```diff
@@ -0,0 +1,144 @@
+# ADR 003: Migrate to jellyfish merkle trie
+
+## Changelog
+* 31-03-2020: Initial Draft
+
+## Context
+
+Need a merkle trie library to store staking state, current one has several disadvantages:
+
+- Written and maintained by single person.
+- Custom data structure and algorithm.
+- Not enough documentation.
+
+## Decision
+
+### jellyfish merkle trie
+
+https://github.com/libra/libra/blob/master/storage/jellyfish-merkle/src/lib.rs
+
+#### Advantages
+
+- Review from Facebook security engineers
+- Small code base, well documented.
+- Well known data structure (simplified version of ethereum MPT).
+- Easy to integrate.
+- Optimized for pbft: Use u64 version number instead of hash.
+- License is apache 2.0.
+- Support inclusion proof and range proof.
+
+#### Disadvantages
+
+- Tied with a lot of custom Libra crates, need to extract an independent library.
+
+  Initial work has done here: https://github.com/yihuang/libra
+
+- No extension node compared to ethereum MPT, deeper tree in theory, a trade off made in favor of simplicity.
```

Comment:
this is not a "decision" -- this is like a context mixed with consequences ?

Reply:
The whole section of "Disadvantages" can be moved to consequences then.

Reply:
Done